dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/scheduling/ContinuousAnalyticsTableJob.java

```diff
@@ -0,0 +1,134 @@
+import org.springframework.stereotype.Component;
+
+import static org.hisp.dhis.util.DateUtils.getMediumDateString;
+
+/**
+ * Job for continuous update of analytics tables. Performs analytics table update on a schedule
+ * where the full analytics table update is done once per day, and the latest analytics partition
+ * update is done with a fixed delay.
+ * <p>
+ * When to run the full update is determined by {@link ContinuousAnalyticsJobParameters#getHourOfDay()},
+ * which specifies the hour of day to run the full update. The next scheduled full analytics table
+ * update time is persisted using a system setting. A full analytics table update is performed
+ * when the current time is after the next scheduled full update time. Otherwise, a partial
+ * update of the latest analytics partition table is performed.
+ *
+ * @author Lars Helge Overland
+ */
+@Component( "continuousAnalyticsTableJob" )
+public class ContinuousAnalyticsTableJob
+    extends AbstractJob
+{
+    private static final Log log = LogFactory.getLog( ContinuousAnalyticsTableJob.class );
+
+    private static final Integer DEFAULT_HOUR_OF_DAY = 0;
+
+    private final AnalyticsTableGenerator analyticsTableGenerator;
+
+    private final SystemSettingManager systemSettingManager;
+
+    public ContinuousAnalyticsTableJob( AnalyticsTableGenerator analyticsTableGenerator, SystemSettingManager systemSettingManager )
+    {
+        this.analyticsTableGenerator = analyticsTableGenerator;
+        this.systemSettingManager = systemSettingManager;
+    }
+
+    @Override
+    public JobType getJobType()
+    {
+        return JobType.CONTINUOUS_ANALYTICS_TABLE;
+    }
+
+    @Override
+    public void execute( JobConfiguration jobConfiguration )
+    {
+        ContinuousAnalyticsJobParameters parameters = (ContinuousAnalyticsJobParameters) jobConfiguration.getJobParameters();
+
+        Integer hourOfDay = ObjectUtils.firstNonNull( parameters.getHourOfDay(), DEFAULT_HOUR_OF_DAY );
+
+        Date now = new Date();
+        Date defaultNextFullUpdate = DateUtils.getNextDate( hourOfDay, now );
+        Date nextFullUpdate = (Date) systemSettingManager.getSystemSetting( SettingKey.NEXT_ANALYTICS_TABLE_UPDATE, defaultNextFullUpdate );
+
+        if ( now.after( nextFullUpdate ) )
+        {
+            log.info( String.format(
+                "Performing full analytics table update, current time: %s, next full update was: %s",
+                getMediumDateString( now ), getMediumDateString( nextFullUpdate ) ) );
+
+            AnalyticsTableUpdateParams params = AnalyticsTableUpdateParams.newBuilder()
+                .withLastYears( parameters.getLastYears() )
+                .withSkipResourceTables( false )
+                .withSkipTableTypes( parameters.getSkipTableTypes() )
+                .withJobId( jobConfiguration )
+                .withStartTime( now )
+                .build();
+
+            analyticsTableGenerator.generateTables( params );
```

Comment:
If an exception is thrown here, the `update` variable will not be updated and, I think, the scheduler will stop working?

Reply:
Good point. Added a try-finally clause.